### ISSUE TITLE
small inversion question

### ISSUE THREAD

**Issue by zhen0z** [modichirag/GSM-VI#5]
Hi Chirag,

I’m currently using your GSM method to solve a linear inversion problem, but the results I’m getting from GSM seem quite off. Could you offer some guidance on what might be going wrong? Below is the code and the results I’ve observed. Thank you for your help!

The results first:
True mean :  [0.74666673 0.92583352 0.22908878 0.25395782]
Fit gsm  :  [11.80972591 31.24928018 24.86131934 -5.42401872]
Fit advi  :  [0.86637243 0.85332108 0.29256509 0.22010893]

import numpy as np
import matplotlib.pyplot as plt
import jax.numpy as jnp
from jax import jit, grad, random
import numpyro.distributions as dist
import optax
import os, sys
sys.path.append(os.getcwd())

# enable 16 bit precision for jax required for lbfgs initializer
from jax import config
config.update("jax_enable_x64", True)

from gsmvi.gsm import GSM
from gsmvi.advi import ADVI
from gsmvi.initializers import lbfgs_init
from gsmvi.monitors import KLMonitor

def setup_model(D):
    # Setup a Gaussian target distribution
    mean = np.random.random(D)
    y_obs = 2 * mean
    L = np.random.normal(size=D**2).reshape(D, D)
    cov = np.matmul(L, L.T) + np.eye(D) * 1e-3
    icov = np.linalg.inv(cov)

    # Functions for log_prob and gradient, supplied by the user
    def lp(x):
        x = x.reshape(-1, D)  # Ensure x is 2D (with shape [batch_size, D])
        sigma2 = 1.0  # Assumed variance
        lp = 0
        for i in range(x.shape[0]):
            lp += -0.5 * np.sum((y_obs - 2 * x[i]) ** 2) / sigma2
        return jnp.array(lp)

    def lp_g(x):
        x = x.reshape(-1, D)  # Ensure x is 2D (with shape [batch_size, D])
        sigma2 = 1.0  # Assumed variance
        lp_g = []
        for i in range(x.shape[0]):
            lp_g.append((y_obs - 2 * x[i]) * (-2) / sigma2)
        return jnp.array(lp_g)
    return mean, cov, lp, lp_g

def gsm_fit(D, lp, lp_g, mean_init, cov_init, lbfgs_res):
    print("Now fit with GSM")
    niter = 500
    batch_size = 1
    key = random.PRNGKey(99)
    monitor = KLMonitor(batch_size_kl=32, checkpoint=10, \
                        offset_evals=lbfgs_res.nfev) #note the offset number of evals

    gsm = GSM(D=D, lp=lp, lp_g=lp_g)
    mean_fit, cov_fit = gsm.fit(key, mean=mean_init, cov=cov_init, niter=niter, batch_size=batch_size, monitor=monitor)
    return mean_fit, cov_fit, monitor

#
def advi_fit(D, lp, lp_g, mean_init, cov_init, lbfgs_res):
    print("\nNow fit with ADVI")
    niter = 500
    lr = 1e-2
    batch_size = 1
    key = random.PRNGKey(99)
    opt = optax.adam(learning_rate=lr)
    monitor = KLMonitor(batch_size_kl=32, checkpoint=10, \
                        offset_evals=lbfgs_res.nfev) #note the offset number of evals

    advi = ADVI(D=D, lp=lp)
    mean_fit, cov_fit, losses = advi.fit(key, mean=mean_init, cov=cov_init, opt=opt, batch_size=batch_size, niter=niter, monitor=monitor)
    return mean_fit, cov_fit, monitor

if __name__=="__main__":
    
    ###
    # setup a toy Gaussia model and extracet score needed for GSM
    D = 4
    mean, cov, lp_fn, lp_g_fn = setup_model(D=D)
    lp = jit(lambda x: lp_fn(x))  # Ensure lp_fn uses JAX
    lp_g = jit(lambda x: lp_g_fn(x))  # Ensure lp_g_fn uses JAX

    ###
    print("Initialize with LBFGS")
    mean_init = np.ones(D)     # setup gsm with initilization from LBFGS fit
    mean_init, cov_init, lbfgs_res = lbfgs_init(mean_init, lp, lp_g)
    print(f'LBFGS fit: \n{lbfgs_res}\n')
    
    mean_gsm, cov_gsm, monitor_gsm = gsm_fit(D, lp, lp_g, mean_init, cov_init, lbfgs_res)
    mean_advi, cov_advi, monitor_advi = advi_fit(D, lp, lp_g, mean_init, cov_init, lbfgs_res)

    # Check that the output is correct
    print("\nTrue mean : ", mean)
    print("Fit gsm  : ", mean_gsm)
    print("Fit advi  : ", mean_advi)

    # Check that the KL divergence decreases
    plt.plot(monitor_gsm.nevals, monitor_gsm.rkl, label='GSM')
    plt.plot(monitor_advi.nevals, monitor_advi.rkl, label='ADVI')
    plt.legend()
    plt.xlabel("Iteration")
    plt.ylabel("Reverse KL")
    plt.savefig("monitor_kl.png")
